2.TCGA-MLexample-covariates.py

```diff
@@ -0,0 +1,355 @@
+
+
+# In[16]:
+
+# Best Params
+for model, pipeline in cv_pipelines.items():
+    print('{0}: {1:.3%}'.format(model, pipeline.best_score_))
+
+    # Best Params
+    print(pipeline.best_params_)
+
+
+# ## Visualize hyperparameters performance
+
+# In[17]:
+
+cv_results_df_dict = {model: 
+    pd.concat([
+        pd.DataFrame(pipeline.cv_results_),
+        pd.DataFrame.from_records(pipeline.cv_results_['params']),
+    ], axis='columns') for model, pipeline in cv_pipelines.items()}
+
+model = 'full'
+
+cv_results_df_dict[model].head(2)
+
+
+# In[18]:
+
+# Cross-validated performance heatmap
+model = 'full'
+
+cv_score_mat = pd.pivot_table(cv_results_df_dict[model],
+                              values='mean_test_score', 
+                              index='classify__l1_ratio',
+                              columns='classify__alpha')
+ax = sns.heatmap(cv_score_mat, annot=True, fmt='.1%')
+ax.set_xlabel('Regularization strength multiplier (alpha)')
+ax.set_ylabel('Elastic net mixing parameter (l1_ratio)');
+
+
+# ## Use Optimal Hyperparameters to Output ROC Curve
+
+# In[19]:
+
+y_pred_dict = {
+    model: {
+        'train': pipeline.decision_function(X_partitions[model]['train']),
+        'test':  pipeline.decision_function(X_partitions[model]['test'])
+    } for model, pipeline in cv_pipelines.items()
+}
+
+def get_threshold_metrics(y_true, y_pred):
+    roc_columns = ['fpr', 'tpr', 'threshold']
+    roc_items = zip(roc_columns, roc_curve(y_true, y_pred))
+    roc_df = pd.DataFrame.from_items(roc_items)
+    auroc = roc_auc_score(y_true, y_pred)
+    return {'auroc': auroc, 'roc_df': roc_df}
+
+metrics_dict = {    
+    model: {
+        'train': get_threshold_metrics(y_train, y_pred_dict[model]['train']),
+        'test':  get_threshold_metrics(y_test, y_pred_dict[model]['test'])
+    } for model in y_pred_dict.keys()
+}
+
+
+# In[20]:
+
+# TODO: do not save intermediate files?
+# Assemble the data for ROC curves
+model_order = ['full', 'expressions', 'covariates']
+
+auc_output = pd.DataFrame()
+roc_output = pd.DataFrame()
+
+for model in model_order:
+    metrics_partition = metrics_dict[model]
+    for partition, metrics in metrics_partition.items():
+        auc_output = auc_output.append(pd.DataFrame({
+            'partition': [partition],
+            'feature_set': [model],
+            'auc': metrics['auroc']
+        }))
+        roc_df = metrics['roc_df']
+        roc_output = roc_output.append(pd.DataFrame({
+            'false_positive_rate': roc_df.fpr,
+            'true_positive_rate': roc_df.tpr,
+            'partition': partition,
+            'feature_set': model
+        }))
+auc_output['legend_index'] = range(len(auc_output.index))
+
+roc_output.to_csv('jupyter_data/roc_output.csv', index = False)
+auc_output.to_csv('jupyter_data/auc.csv', index = False)
+
+with open('jupyter_data/roc_vega_spec.json', 'r') as fp:
+    vega_spec = json.load(fp)
+    
+vega.Vega(vega_spec)
```

Comment:
I was hoping you could directly pass the dataframe to `vega.Vega` via the `data` argument. Looking at the [source code](https://github.com/vega/ipyvega/blob/6b0571ee66117247f94f10afcc7ccf1dd8661ed2/vega/base.py#L20-L23), I'm not sure whether `data` does anything, but worth a try. Did you try?

I see above "TODO: do not save intermediate files?"